src/Arcus.Security.Tests.Unit/KeyVault/KeyVaultSecretProviderTests.cs

```diff
@@ -25,38 +31,71 @@ public void KeyVaultSecretProvider_CreateWithHttpScheme_ShouldFailWithUriFormatE
             string uri = $"http://{Guid.NewGuid():N}.vault.azure.net/";
 
             // Act & Assert
-            Assert.ThrowsAny<UriFormatException>(() => new KeyVaultSecretProvider(null, new KeyVaultConfiguration(uri)));
+            Assert.ThrowsAny<UriFormatException>(() => new KeyVaultSecretProvider((IKeyVaultAuthentication) null, new KeyVaultConfiguration(uri)));
         }
 
         [Fact]
         public void KeyVaultSecretProvider_CreateWithoutUri_ShouldFailWithArgumentException()
         {
-            // Arrange
-            string uri = null;
-
             // Act & Assert
-            Assert.ThrowsAny<ArgumentException>(() => new KeyVaultSecretProvider(new AzureKeyVaultAuthenticatorDummy(), new KeyVaultConfiguration(uri)));
+            Assert.ThrowsAny<ArgumentException>(
+                () => new KeyVaultSecretProvider(
+                    new AzureKeyVaultAuthenticatorDummy(), 
+                    new KeyVaultConfiguration(rawVaultUri: null)));
         }
 
         [Fact]
         public void KeyVaultSecretProvider_CreateWithoutClientFactory_ShouldFailWithArgumentException()
         {
             // Arrange
-            string uri = $"https://{Guid.NewGuid():N}.vault.azure.net/";
+            string uri = GenerateVaultUri();
 
             // Act & Assert
-            Assert.ThrowsAny<ArgumentException>(() => new KeyVaultSecretProvider(null, new KeyVaultConfiguration(uri)));
+            Assert.ThrowsAny<ArgumentException>(() => new KeyVaultSecretProvider((IKeyVaultAuthentication) null, new KeyVaultConfiguration(uri)));
         }
 
         [Fact]
         public void KeyVaultSecretProvider_CreateWithValidArguments_ShouldSucceed()
         {
             // Arrange
-            string uri = $"https://{Guid.NewGuid():N}.vault.azure.net/";
+            string uri = GenerateVaultUri();
 
             // Act & Assert
             var secretProvider = new KeyVaultSecretProvider(new AzureKeyVaultAuthenticatorDummy(), new KeyVaultConfiguration(uri));
             Assert.NotNull(secretProvider);
         }
+
+        [Fact]
+        public async Task KeyVaultSecretProvider_GetsSecretValue_AfterRetriedTooManyRequestException()
+        {
+            // Arrange
+            string secretName = $"secret-name-{Guid.NewGuid()}";
+            string expected = $"secret-value-{Guid.NewGuid()}";
+
+            var keyVaultClient = new SimulatedKeyVaultClient(
+                request => throw new KeyVaultErrorException("Sabotage secret retrieval with TooManyRequests")
```

Comment:
It must be my ignorance, but this is very strange syntax

Reply:
Strange in what way?

Reply:
1. A lot of stuff which is done in constructor call
2. THrow exception after which we run a lambda
3. Request not being used

![image](https://user-images.githubusercontent.com/4345663/57059960-d6cbb480-6cb7-11e9-9d99-737ca14ce179.png)

Reply:
It's actually a series of simulations (functions) that we should run in sequence for each request.
It's a very common way to use it this way but with the right suggestions I can make it more OOP?

Reply:
You're throwing an exception after which you run a function, not really sure why we are using this kind of syntax, this feels strange.

Reply:
1. Not entirely correct. The ctor gets a list of simulations which it runs in sequence. The functions are therfore "lazy" executed.
2. This syntax allows the test reader to see the order in which the key vault client will act on incoming requests.
3. It may feel strange because this way of working is using functions as first class citizens just like any other value; which is the core principle of functional programming. But from all syntaxes I've seen to test this; this feels to me as the most readable and maintainable solution.

Reply:
But we are not using functional programming on this project so it does not make sense to shift to that flavor imo.

Anyway, business-wise it's good enough and don't want to keep this PR open any longer than needed so I'm resolving this.

Reply:
:D That's funny you say that. LINQ, async/await, delegates/lamdba, ... are all functional concepts. So you should't be using those constructs either. If you're using a recent version of C#, you're almost always using some functional programming ideas.